centerofci/mathesar
# Shareable links frontend - shared table consumer view

Related to #3036 

**This PR:**
* Allows viewing shared tables on the frontend

**Testing instructions:**
* Generate a shared link for a table via API:
  - A post request to `/api/ui/v0/tables/1/shares` with `{ enabled: true }` should share the table and generate a slug. The response will be as follows:
    ```
    {
      "id": 1,
      "slug": "874cfd2b-d580-4ea1-9ded-c226d46f9443",
      "enabled": true
    }
    ```
* Visit `http://localhost:8000/shares/tables/874cfd2b-d580-4ea1-9ded-c226d46f9443/` and you should be able to see a readonly view of the table.
* The following cases should be taken into account, when the link is valid:
  - a. The link should work when you're not logged into Mathesar.
  - b. The link should work when the user is logged in but does not have access to the table's schema or DB.
  - c. The link should work when the user is logged in and has access to the table.

**Note for reviewers:**
* Case 'b' mentioned above was the trickiest to implement, please pay extra attention to it.
* Inorder to make 'b' more stable, a complete store refactor is required on the frontend. In this PR, a workaround has been implemented for the time being. Some examples on why a store refactor is required:
   - For case 'b', code like this `$tables.get(id)` will not work because the user does not have access to the table, and therefore `tables` store will not contain the requested table.
    - Imports like `$currentDbAbstractTypes` will return empty values because the user does not have access to the database.
* The workaround implemented here is to separate the routing contexts of shared routes and normal authenticated routes. This effectively makes the routing of shared routes server-side-only.
* This is more of a band-aid solution and we should get into refactoring the stores as soon as possible in upcoming cycles.

**Cases not handled in this PR:**
* _Filtering FK columns_, and _context menu links_ do not work because the user does not have access to these entities. They are currently visible on the UI. We'll have to come up with a UX and have some discussions around how to handle these cases. 


## Checklist
 
  index.md`).
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
